tests/proxyname_test.go

```diff
@@ -78,55 +82,70 @@ func TestProxyNameLabel(t *testing.T) { // nolint
 			return ctx
-		}).
-		Teardown(func(context.Context, *testing.T, *envconf.Config) context.Context {
-			services := []*kubernetes.Service{disabledService.(*kubernetes.Service), toggledService.(*kubernetes.Service)}
-			tools.ResetTestBoard(t, services, model)
-			return ctx
-		}).
-		Assess("should implement service.kubernetes.io/service-proxy-name", func(_ context.Context, t *testing.T, cfg *envconf.Config) context.Context {
-			ma.Logger.Info("verify the toggledServices are up.")
+		}
+	}
+
+	teardown := func(context.Context, *testing.T, *envconf.Config) context.Context {
+		for _, service := range []kubernetes.ServiceBase{disabledService, toggledService} {
+			if err := service.Delete(); err != nil {
+				t.Fatal(err)
+			}
+		}
+		return ctx
+	}
+
+	getAssessFunc := func(labelKey string) features.Func {
+		return func(_ context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			ma.Logger.Info("verify the toggledService is up without", zap.String("label", labelKey))
 			toPod.SetClusterIP(toggledClusterIP)
 			tools.MustNoWrong(matrix.ValidateOrFail(ma, model, &matrix.TestCase{
 				ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: upReachability,
 				ServiceType: entities.ClusterIP,
 			}, false), t)
 
-			ma.Logger.Info("verify the disabledServices are not up.")
+			ma.Logger.Info("verify the disabledService is not up with", zap.String("label", labelKey))
 			toPod.SetClusterIP(disabledClusterIP)
 			tools.MustNoWrong(matrix.ValidateOrFail(ma, model, &matrix.TestCase{
 				ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: downReachability,
 				ServiceType: entities.ClusterIP,
 			}, false), t)
 
-			ma.Logger.Info("add service-proxy-name label.")
-			mustOrFatal(toggledService.SetLabel(labelKey, labelValue), t)
+			ma.Logger.Info("add label to the toggledService", zap.String("label", labelKey))
+			mustOrFatal(toggledService.SetLabel(proxyNameLabelKey, labelValue), t)
 
-			ma.Logger.Info("verify the toggledServices are not up.")
+			ma.Logger.Info("verify the toggledService is not up with", zap.String("label", labelKey))
 			toPod.SetClusterIP(toggledClusterIP)
 			tools.MustNoWrong(matrix.ValidateOrFail(ma, model, &matrix.TestCase{
 				ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: downReachability,
 				ServiceType: entities.ClusterIP,
 			}, false), t)
 
-			ma.Logger.Info("remove service-proxy-name label.")
-			mustOrFatal(toggledService.RemoveLabel(labelKey), t)
+			ma.Logger.Info("remove label from the toggledService", zap.String("label", labelKey))
+			mustOrFatal(toggledService.RemoveLabel(proxyNameLabelKey), t)
 
-			ma.Logger.Info("verify the toggledServices are up again.")
+			ma.Logger.Info("verify the toggledService is up again without", zap.String("label", labelKey))
 			tools.MustNoWrong(matrix.ValidateOrFail(ma, model, &matrix.TestCase{
 				ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: upReachability,
 				ServiceType: entities.ClusterIP,
 			}, false), t)
```

Comment:
it can be trickier because headless, is not accessible with the ClusterIP, but still works with the DNS (should resolve directly to the pod), what are we testing here?

1. Not reachable at all on ClusterIP
2. Works with DNS

Reply:
agree^, looks like there should be no cluster IP for headless service, and need to use the service name here.

Reply:
(1)

here we are only asserting it is not accessible through ClusterIP  if that service were labeled. not relatable to DNS for this test.

Reply:
oh the third argument for `ValidateOrFail` is `ignoreLoopback`, here we are asserting the service is reachable with `upReachability`

Reply:
for this test case only, it is testing readability of svc's ClusterIP, once the `headless` tag is removed